src/error/error.h

```diff
@@ -0,0 +1,199 @@
+
+  /// Return the error message it was constructed with.
+  virtual std::string toString() const override { return message(); }
+
+  virtual ~GenericError() {}
+ protected:
+  /// Access to the field containing the error message passed to the
+  /// constructor.
+  std::string message() const { return message_; }
+ private:
+  const std::string message_;
+};
+
+
+/// Represents either an error or a T.
+/// This class is immutable (but the T can be mutable).
+template <typename T, typename Err = GenericError>
+class ErrorOr {
+  static_assert(std::is_base_of<Error, Err>::value,
+                "Error type must be a subclass of Error");
+ public:
+  // Constructors from value or errors.
+
+  // Not explicit on purpose.
+  ErrorOr(const Err& error) : union_(error), is_ok_(false) {}
+
+  // Not explicit on purpose.
+  ErrorOr(Err&& error) : union_(std::move(error)), is_ok_(false) {}
+
+  // Not explicit on purpose.
+  ErrorOr(const T& value) : union_(value), is_ok_(true) {}
+
+  // Not explicit on purpose.
+  ErrorOr(T&& value) : union_(std::move(value)), is_ok_(true) {}
+
+  // Copy constructor.
+  ErrorOr(const ErrorOr& other)
+      : is_ok_(other.is_ok_)
+  {
+    if (is_ok_)
+      new (&union_.value) T(other.union_.value);
+    else
+      new (&union_.error) Err(other.union_.error);
+  }
+
+  // Move constructor.
+  ErrorOr(ErrorOr&& other)
+      : is_ok_(other.is_ok_)
+  {
+    if (is_ok_)
+      new (&union_.value) T(std::move(other.union_.value));
+    else
+      new (&union_.error) Err(std::move(other.union_.error));
+  }
+
+  // Copy assignment.
+  ErrorOr& operator=(const ErrorOr& other) {
+    this->~ErrorOr();
+    is_ok_ = other.is_ok_;
+    if (is_ok_)
+      new (&union_.value) T(other.union_.value);
+    else
+      new (&union_.error) Err(other.union_.error);
+    return *this;
+  }
+
+  // Move assignment.
+  ErrorOr& operator=(ErrorOr&& other) {
+    this->~ErrorOr();
+    is_ok_ = other.is_ok_;
+    if (is_ok_)
+      new (&union_.value) T(std::move(other.union_.value));
+    else
+      new (&union_.error) Err(std::move(other.union_.error));
+    return *this;
+  }
+
+  /// Check whether it is an error or a value.
+  bool is_ok() const { return is_ok_; }
+
+  /// Return the value if it is one, fail otherwise.
+  T& value_or_die() {
+    if (!is_ok_)
+      throw std::domain_error("ErrorOr was error, asked for value");
+    return union_.value;
+  }
+
+  /// Return the value if it is one, fail otherwise.
+  const T& value_or_die() const {
+    if (!is_ok_)
+      throw std::domain_error("ErrorOr was error, asked for value");
+    return union_.value;
+  }
+
+  /// Return the error if it is one, fail otherwise.
+  const Err& error_or_die() const {
+    if (is_ok_)
+      throw std::domain_error("ErrorOr was value, asked for error");
+    return union_.error;
+  }
+
+  /// Return the error if it is one, otherwise return "Ok".
+  std::string toString() const {
+    if (is_ok_)
+      return "Ok";
```

Comment:
If the value can be displayed, it might be interesting to actually display it instead of "Ok". This can be done using tag dispatching, but you don't need to do this in the current PR.